### ISSUE TITLE
Support for custom `error` catcher for api

### ISSUE THREAD

**Issue by nechaido** [metarhia/impress#1812]
### Describe the problem

It is often not enough to catch any error that arises in the application code and to send error 500 to the client.

### Describe the solution

Better solution would be to allow custom API handler wrappers to be provided

### Additional context

See https://github.com/metatech-university/NodeJS-Application/pull/4/files#diff-76374afb32b513f9f1f3ad04638ffcfba3c4cd36aa447236a128c2b6ed6c0936R14

**Comment by tshemsedinov** [comment]
Concerning errors: I think we can use single Error class with following features:
`new Error(message: string, { code: number, cause: Error })`

**Comment by mprudnik** [comment]
Another option:

```javascript
class DomainError extends Error {
...
}

class MessengerError extends DomainError {
...

  static invalidPermission() {
    return new MessengerError('Permission denied');
  }
  
  static notFound() {
    return new MessengerError('Not found');
  }
  
  // and so on...
}
```
This neither requires defining a class per each error (1 proposal) nor requires additional functionality to check if a general purpose error received an allowed message (2 proposal).
This also adds static checks and autocompletion.

**Comment by georgolden** [comment]
I suggest to have `onError` handler for API method like this:

```js
({
  parameters: {
    a: 'number',
    b: 'number',
  },

  method: async ({ a, b }) => {
    const result = a + b;
    return result;
  },

  onError: (error) =>
    'code' in error ? error : { message: error.message, code: 500 },

  returns: 'number',
});

```

The goal is to have any custom error handlers in case if API method failed.

Inspired from Fastify hooks 
https://www.fastify.io/docs/latest/Reference/Hooks/#onerror
https://www.fastify.io/docs/latest/Reference/Lifecycle/

**Comment by DemianParkhomenko** [comment]
It is just a draft proposal/example of how we can declare a list of the errors with TypeScript. `data: any` can be typed better. 

```ts
export enum ErrorCode {
  notFound = 'NOT_FOUND',
  invalidResponse = 'INVALID_RESPONSE',
  authenticationError = 'AUTHENTICATION_ERROR',
  networkError = 'NETWORK_ERROR',
}

const list = new Map<ErrorCode, { message: string; data?: any }>([
  [ErrorCode.notFound, { message: 'Not found' }],
  [
    ErrorCode.invalidResponse,
    {
      message: 'Invalid response. Please try again later.',
      data: { additionalFields: 'something' },
    },
  ],
  [
    ErrorCode.authenticationError,
    {
      message: 'Authentication failed',
      data: { example: 401, something: 'something' },
    },
  ],
  [
    ErrorCode.networkError,
    {
      message: 'Network error occurred',
      data: { example: 500 },
    },
  ],
]);

export const createError = (code: ErrorCode) => ({ code, ...list.get(code) });

console.log(createError(ErrorCode.notFound));

for (const v of list.values()) console.log(v);

```

**Comment by Haliont** [comment]
```
// error names
const DOMAIN_ERROR_NAME = {
  NotFound = 'NotFound',
  Forbidden = 'Forbidden',
  SomeEntityIsDeactivated = 'SomeEntityIsDeactivated',
  ...
};

// domain layer
throw new DomainError(DOMAIN_ERROR_NAME.NotFound)
```

**Comment by lundibundi** [comment]
```ts
export interface LogFormatError {
  toJSON(): Record<string, unknown>;

  // compatible with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior.
}

export interface UserFormatError {
  toUserError(): Record<string, unknown>;

  // allows to provide custom data when error is sent to user
}
```

both interfaces are optional for all Errors. Logger/API will check for those upon handling.

**Comment by tshemsedinov** [comment]
Endpoint code:
```js
({
  parameters: {
    person: { domain: 'Person' },
    address: { domain: 'Address' },
  },

  method: async ({ person, address }) => {
    const addressId = await api.gs.create(address);
    person.address = addressId;
    const personId = await api.gs.create(person);
    return personId;
  },

  returns: { type: 'number' },

  errors: {
    ECONTRACT: 'Invalid arguments',
    ECANTSAVE: 'Person and address can not be saved to database',
    ERESULT: 'Invalid result',
  },
});
```
We can generate `example.d.ts` automaticaly
```ts
type Code = 'ECONTRACT' | 'ECANTSAVE' |  'ERESULT';
class CustomError {
  constructor(message: string, options: { code: Code });
}
```
![Screenshot from 2023-06-23 04-31-56](https://github.com/metarhia/impress/assets/4405297/686a0ee0-0e03-47a8-abff-4c927e264ac5)

**Comment by tshemsedinov** [comment]
I believe, we need complex examples, not just a code snippet, we need branch/fork with a few files to show how it will work and ready to try branch, for example my proposal: https://github.com/metarhia/Example/commit/ef7f4833d12f3030c4d40a0ea14e13ca549ba4d1
@lundibundi @nechaido @mprudnik @georgolden @DemianParkhomenko @Haliont

**Comment by mprudnik** [comment]
Added my extended proposal - https://github.com/metarhia/Example/pull/234.
Additional notes regarding error hooks:
I propose to have multiple places, where `onError` handler can be defined - in method itself (near handler) and `error.js` file on service level. Then we can trigger the nearest one in case of error:
- if method defines own `onError` - use it ignore service and global error handler
- else if service defines own `onError` - use it and ignore global error handler
- else use global error handler.

Also, I'm not sure if those handlers should be triggered if `error` is an instance of `DomainError` since normally that means that it is part of expected behavior and the error can be serialized and sent to the client.

**Comment by tshemsedinov** [comment]
Implemented in Version 3.0.0
